nucleus/admin/server-mgmt/src/main/java/com/sun/enterprise/admin/servermgmt/KeystoreManager.java

```diff
@@ -299,27 +310,96 @@ protected void copyCertificates(File keyStore, File trustStore, DomainConfig con
 
-        String[] keytoolCmd = {
-            "-importkeystore",
-            "-srckeystore", javaTrustStore.getAbsolutePath(),
-            "-destkeystore", trustStore.getAbsolutePath(),
-            "-noprompt"
-        };
+        // Load the java trust store
+        KeyStore javaTrustStore;
+        KeyStore destTrustStore;
+        try {
+            FileInputStream javaIn = new FileInputStream(javaTrustStoreFile);
+            javaTrustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+            javaTrustStore.load(javaIn, "changeit".toCharArray());
+            javaIn.close();
+            
+            FileInputStream destIn = new FileInputStream(trustStore);
+            destTrustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+            destTrustStore.load(destIn, masterPassword.toCharArray());
+            destIn.close();
+        } catch (KeyStoreException ex) {
+            throw new RepositoryException("Unable to get Keystore instance.", ex);
+        } catch (NoSuchAlgorithmException ex) {
+            throw new RepositoryException("Unable to read Keystore instance.", ex);
+        } catch (CertificateException ex) {
+            throw new RepositoryException("Unable to load certificate from Keystore instance.", ex);
+        } catch (FileNotFoundException ex) {
+            throw new RepositoryException("Unable to find Keystore instance.", ex);
+        } catch (IOException ex) {
+            throw new RepositoryException("Unexpected exception reading Keystore instance.", ex);
+        }
+
+        // Load the valid certificates to the store
+        Map<String, Certificate> validCerts = getValidCertificateAliases(javaTrustStore, "changeit");
+        try {
+            for(String alias : validCerts.keySet()) {
+                Certificate cert = validCerts.get(alias);
+                if (!destTrustStore.containsAlias(alias)) {
+                    destTrustStore.setCertificateEntry(alias, cert);
+                }
+            }
+        } catch (KeyStoreException ex) {
+            throw new RepositoryException("Keystore hasn't been initialized.", ex);
+        }
+
+        // Load the store back to the initial file
+        try {
+            FileOutputStream out = new FileOutputStream(trustStore);
+            destTrustStore.store(out, masterPassword.toCharArray());
+            out.flush();
+            out.close();
+	} catch (KeyStoreException | NoSuchAlgorithmException | CertificateException | IOException ex) {
+            throw new RepositoryException("Unexpected exception writing certificates to the Keystore instance.", ex);
+        }
+
+    }
+
+    protected Map<String, Certificate> getValidCertificateAliases(KeyStore keyStore, String keyStorePassword)
+            throws RepositoryException {
+
+        Map<String, Certificate> validCerts = new HashMap<>();
+
+        try {
+            Enumeration<String> aliases = keyStore.aliases();
```

Comment:
Perhaps the following is a tad more clear/modern:

```java
for (String alias : list(keyStore.aliases()) {
    // ...
}
```

With `list` statically imported from [Collections.list](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#list-java.util.Enumeration-)